src/Features/Core/Portable/MetadataAsSource/FileSystemMetadataDocumentPersister.cs

```diff
@@ -0,0 +1,169 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.MetadataAsSource;
+
+internal sealed class FileSystemMetadataDocumentPersister : IMetadataDocumentPersister
+{
+    /// <summary>
+    /// We create a mutex so other processes can see if our directory is still alive.  As long as we own the mutex, no
+    /// other VS instance will try to delete our _rootTemporaryPathWithGuid folder.
+    /// </summary>
+#pragma warning disable IDE0052 // Remove unread private members
+    private readonly Mutex _mutex;
+#pragma warning restore IDE0052 // Remove unread private members
+    private readonly string _rootTemporaryPathWithGuid;
+    private readonly string _rootTemporaryPath = Path.Combine(Path.GetTempPath(), MetadataAsSourceFileService.MetadataAsSource);
+
+    public FileSystemMetadataDocumentPersister()
+    {
+        var guidString = Guid.NewGuid().ToString("N");
+        _rootTemporaryPathWithGuid = Path.Combine(_rootTemporaryPath, guidString);
+        _mutex = new Mutex(initiallyOwned: true, name: CreateMutexName(guidString));
+    }
+
+    private static string CreateMutexName(string directoryName)
+        => $"{MetadataAsSourceFileService.MetadataAsSource}-{directoryName}";
+
+    public bool TryGetExistingText(string documentPath, Encoding encoding, SourceHashAlgorithm checksumAlgorithm, Func<SourceText, bool> verifyExistingDocument, [NotNullWhen(true)] out SourceText? sourceText)
+    {
+        sourceText = null;
+        if (!File.Exists(documentPath))
+            return false;
+
+        using var stream = new FileStream(documentPath, FileMode.Open, FileAccess.Read, FileShare.Read | FileShare.Delete);
+
+        sourceText = SourceText.From(stream, encoding, checksumAlgorithm, throwIfBinaryDetected: true);
+
+        if (verifyExistingDocument is not null && !verifyExistingDocument(sourceText))
+            return false;
+
+        return true;
+    }
+
+    public async Task<bool> WriteMetadataDocumentAsync(string documentPath, Encoding encoding, SourceText text, Action<Exception>? logFailure, CancellationToken cancellationToken)
+    {
+        try
+        {
+            var directoryName = Path.GetDirectoryName(documentPath)!;
+            if (!await EnsureDirectoryExistsAsync(directoryName, cancellationToken).ConfigureAwait(false))
+                throw new IOException($"Unable to create directory: {directoryName}");
```

Comment:
If we're just going to throw an exception, should we have EnsureDirectoryExistsAsync() just throw it? It seems a bit odd that we have PerformIOs to eat the exception, but we'll then lose the underlying cause here.